libcloudforensics/aws.py

```diff
@@ -103,49 +87,50 @@ def ListInstances(self, region=None, filters=None):
     Returns:
       dict: Dictionary with name and metadata for each instance.
 
+    Raises:
+      RuntimeError: If instances can't be listed.
+
     Example usage:
       ListInstances(region='us-east-1', filters=[
-        dict(Name='instance-id', Values=['a-particular-instance-id'])
-      ])
+          dict(Name='instance-id', Values=['some-instance-id'])])
     """
     if not filters:
       filters = []
 
     instances = dict()
-    have_all_tokens = False
     next_token = None
+    client = self.ClientApi(EC2_SERVICE, region=region)
 
-    while not have_all_tokens:
-      if next_token:
-        response = self.ClientApi(
-            EC2_SERVICE, region=region).describe_instances(
-                Filters=filters,
-                NextToken=next_token)
-      else:
-        response = self.ClientApi(
-            EC2_SERVICE, region=region).describe_instances(
-                Filters=filters)
-
-      for reservation in response["Reservations"]:
-        for instance in reservation["Instances"]:
-          instance_name = None
-          if instance.get('Tags'):
-            for tag in instance['Tags']:
-              instance_name = tag.get('Value') if tag.get('Key') == 'Name' \
-                else None
-              if instance_name is not None:
-                break
+    while True:
+      try:
+        if next_token:
+          response = client.describe_instances(
+              Filters=filters, NextToken=next_token)
+        else:
+          response = client.describe_instances(Filters=filters)
+      except client.exceptions.ClientError as exception:
+        raise RuntimeError('Error: could not retrieve instances: {0:s}'.format(
+            str(exception)))
+
+      for reservation in response['Reservations']:
+        for instance in reservation['Instances']:
           # Terminated instances are filtered out
-          if instance['State']['Name'] != 'terminated':
-            instance_info = dict(region=self.default_region,
-                                 zone=instance['Placement']['AvailabilityZone'])
-            if instance_name:
-              instance_info.update(name=instance_name)
-            instances[instance['InstanceId']] = instance_info
+          if instance['State']['Name'] == 'terminated':
+            continue
+
+          instance_info = dict(region=self.default_region,
+                               zone=instance['Placement']['AvailabilityZone'])
+
+          for tag in instance.get('Tags', []):
```

Comment:
Correct. Since EC2 instances are not referenced to by their names like in GCP but only their IDs, adding a name to any instance is completely optional for users. If none is assigned to an instance, then `instance_info` will only contain region and zone information.

Reply:
Ack, thanks!